sn_node/src/put_validation.rs

```diff
@@ -758,33 +741,109 @@ impl Node {
-            Err(Error::InvalidRequest(format!(
-                "Found no valid spends while validating Spend PUT for {unique_pubkey:?}"
-            )))
-        } else if verified_spends.len() > 1 {
-            warn!("Got a double spend for {unique_pubkey:?}");
-            Ok(verified_spends)
-        } else {
-            debug!("Got a single valid spend for {unique_pubkey:?}");
-            Ok(verified_spends)
+        // return 2 spends max
+        let all_verified_spends: Vec<_> = all_verified_spends.into_iter().collect();
+        match all_verified_spends.as_slice() {
+            [one_spend] => Ok((one_spend.clone(), None)),
+            [one, two] => Ok((one.clone(), Some(two.clone()))),
+            [] => {
+                warn!("Invalid request: none of the spends were valid for {unique_pubkey:?}");
+                Err(Error::InvalidRequest(format!(
+                    "Found no valid spends while validating Spends for {unique_pubkey:?}"
+                )))
+            }
+            more => {
+                warn!("Got more than 2 verified spends, this might be a double spend spam attack, making sure to favour live branches (branches with spent descendants)");
+                let (one, two) = self.verified_spends_select_2_live(more).await?;
+                Ok((one, Some(two)))
+            }
+        }
+    }
+
+    async fn verified_spends_select_2_live(
+        &self,
+        many_spends: &[SignedSpend],
+    ) -> Result<(SignedSpend, SignedSpend)> {
+        // get all spends descendants
+        let mut tasks = JoinSet::new();
+        for spend in many_spends {
+            let descendants: BTreeSet<_> = spend
+                .spend
+                .spent_tx
+                .outputs
+                .iter()
+                .map(|o| o.unique_pubkey())
+                .map(SpendAddress::from_unique_pubkey)
+                .collect();
+            for d in descendants {
+                let self_clone = self.clone();
+                let spend_clone = spend.to_owned();
+                let _ = tasks.spawn(async move {
+                    let res = self_clone.network().get_raw_spends(d).await;
+                    (spend_clone, res)
+                });
+            }
+        }
+
+        // identify up to two live spends (aka spends with spent descendants)
+        let mut live_spends = BTreeSet::new();
+        while let Some(res) = tasks.join_next().await {
+            match res {
+                Ok((spend, Ok(_descendant))) => {
+                    let _inserted = live_spends.insert(spend);
+                    if live_spends.len() > 1 {
+                        let array: Vec<_> = live_spends.clone().into_iter().collect();
+                        if let [one, two] = array.as_slice() {
+                            warn!("Got two live spends {one:?} and {two:?}, things are messed up!");
+                            return Ok((one.to_owned().clone(), two.to_owned().clone()));
```

Comment:
If we return early here, will we have an inconsistent state among our nodes if there are more than 2 live branches? 
Having >1 live branch is very unlikely, but should we collect all live branches and then return the first 2, just to be deterministic?

Reply:
You are absolutely right!